src/main/java/com/metamx/cache/CaffeineCache.java

```diff
@@ -54,58 +55,79 @@ public static CaffeineCache create(final CaffeineCacheConfig config)
           .expireAfterAccess(config.getExpiration(), TimeUnit.MILLISECONDS);
     }
     if (config.getMaxSize() >= 0) {
-      builder = builder
-          .maximumSize(config.getMaxSize())
-          .maximumWeight(config.getMaxSize());
+      builder = builder.maximumWeight(config.getMaxSize());
+      builder.weigher(new Weigher<NamedKey, byte[]>()
+      {
+        @Override
+        public int weigh(@Nonnull NamedKey key, @Nonnull byte[] value)
+        {
+          return value.length;
+        }
+      });
+    }
+
+    final Cache delegateCache = config.getDelegateCache();
+    final CacheLoader<NamedKey, byte[]> cacheLoader;
+    if (delegateCache != null) {
+      cacheLoader = new CacheLoader<NamedKey, byte[]>()
+      {
+        @Override
+        public byte[] load(@Nonnull NamedKey key) throws Exception
+        {
+          return delegateCache.get(key);
+        }
+
+        @Override
+        public Map<NamedKey, byte[]> loadAll(@Nonnull Iterable<? extends NamedKey> keys) throws Exception
+        {
+          final Iterable<NamedKey> keys0 = (Iterable<NamedKey>) keys;
+          return Maps.transformValues(delegateCache.getBulk(keys0), CaffeineCache::serialize);
```

Comment:
You may find [CacheWriter](https://github.com/ben-manes/caffeine/wiki/Writer) useful to compliment the loader. Then all loads and writes are atomic from Caffeine's perspective. Usually races in a look-aside cache are okay, but sometimes they aren't or not worth the effort to reason about.